src/Microsoft.AspNetCore.Server.Kestrel/KestrelServer.cs

```diff
@@ -108,6 +109,10 @@ public void Start<TContext>(IHttpApplication<TContext> application)
                         atLeastOneListener = true;
                         _disposables.Push(engine.CreateServer(
                             parsedAddress));
+
+                        // If requested port was "0", replace with assigned dynamic port.
+                        _serverAddresses.Addresses.Remove(address);
+                        _serverAddresses.Addresses.Add(parsedAddress.ToString());
```

Comment:
I added tests for hostname, IPv4 address, and IPv6 address to `FunctionalTests\AddressRegistrationTests.cs`.  The test currently fails on the IPv6 link-local address which ends with `%18`.  The value returned by context.Request.GetDisplayUrl() is missing the `%18`.  I still need to determine whether this is a test issue or a product issue.